### PR TITLE
[processing] Add API for specifying logical parameter groups

### PR DESCRIPTION
**Pull request by nyalldawson**
Implements a revised version of https://github.com/qgis/QGIS-Enhancement-Proposals/issues/120, taking into account @wonder-sk's more advanced logic requirements.

Initially only a single grass algorithm utilises this (r.cost, for the start coordinate parameters) as a test.

### CODE REVIEW

**Review by alexbruy — APPROVED** [review]
Nice! This will improve UX a lot when we update all relevant algorithms to use parameter groups.

+1 to merge.

**Comment by nyalldawson** [comment]
Thanks @alexbruy !

@wonder-sk Can you confirm that this new API meets your needs too?

**Comment by wonder-sk** [comment]
hi @nyalldawson yeah the API looks good to me.

I must admit I am a bit lost in the logic of checkGroup() - especially how the MaybeSpecified value is meant to work. Would the logic work correctly for the following case?

mutually_exclusive_required( mutually_exclusive_optional(p1, p2), mutually_exclusive_optional(p3, p4) )

If I specify no parameters, this should throw an error. But if I read the code correctly, each sub-group would return MaybeSpecified, giving countSpecified=0 and countMaybeSpecified=2 in the root group, passing the test...

I am trying to understand why do we need MaybeSpecified - my logic says that each term (group) should return true/false saying whether it is present or not, or throw an error if the term's type is not satisfied with sub-groups.

**Comment by nyalldawson** [comment]
@wonder_sk

I tried hard to make it work without the maybe specified use, but it breaks for certain logic without (there's unit tests for these).

But I think in your example you actually want:

    mutually_exclusive_required( mutually_exclusive_required(p1, p2), mutually_exclusive_required(p3, p4) )

Because mutually_exclusive_optional means 0 or 1. So in the example you gave setting only a value for P1 would fail - because both children are satisfied with 0 or 1 Params, so they both are treated as specified and then the test fails the mutually exclusive clause on the parent.

If the children are mutually_exclusive_required then  setting neither means that that child is treated as not-set, so setting only P1 passes the mutually exclusive parent condition. Phew!

File: tests/src/analysis/testqgsprocessing.cpp [hunk 1]
```diff
@@ -177,6 +177,220 @@ class DummyAlgorithm : public QgsProcessingAlgorithm
...
+      childGroup1.type = AllRequired;
+      childGroup1.parameters.insert( "p1" );
+      childGroup1.parameters.insert( "p2" );
+      group.childGroups.append( childGroup1 );
+      addLogicalParameterGroup( group );
+      params.clear();
+
+      QVERIFY( checkParameterValues( params, context, &message ) );
+      params.insert( "p1", true );
+      QVERIFY( !checkParameterValues( params, context, &message ) );
```
Comment (wonder-sk) [inline tests/src/analysis/testqgsprocessing.cpp]: Is this test correct? for `mutually_exclusive_optional( all_required( p1, p2 ) )` I would expect that all four possible combinations are correct ({}, {p1}, {p2}, {p1, p2}).

**Comment by wonder-sk** [comment]
Well I picked the example intentionally that way to dig into possible issues MaybeSpecified ;-)
It is maybe not the most practical example, but it should still work...

I am wondering why the "maybe" is required. Here's how I would try to do the logic:
- a group is "present" if one or more of its terms are present. A parameter is "present" when it is specified.
- for each group:
  - count the number of terms that are "present" (store in variable CP)
  - determine whether the group is valid:
    - all required: CP = total number of terms
    - at least one: CP > 0
    - mutually exclusive required: CP = 1
    - mutually exclusive optional: CP <= 1

File: tests/src/analysis/testqgsprocessing.cpp (hunk 1 above)
Comment (nyalldawson) [reply tests/src/analysis/testqgsprocessing.cpp]: I think so - if you say both P1 AND P2 are required, then specifying only one should be a fail. You either need to leave both missing or specify both.

Comment (wonder-sk) [reply tests/src/analysis/testqgsprocessing.cpp]: But here in fact you are saying that both P1 AND P2 are required in a sub-term... And tehn the top-level `mutually_exclusive_optional( X )` is happy with both outcomes of `all_required( p1, p2 )` (satisfied or not satisfied) ?

**Comment by nyalldawson** [comment]
> It is maybe not the most practical example, but it should still work...

It does work correctly though, just not the way you're interpreting the API. 

The issues arise in that we do have unknown states - anytime a group is optional (e.g mutually exclusive optional). Because in that case not specifying any of the grouped parameters can mean either that you dont want that branch OR that you want that branch, but you want it empty. E.g. the test

   mutually_exclusive_required( mutually_exclusive_optional(P1,P2),  mutually_exclusive_optional(P3,P4))

If neither P1 or P2 are set, we don't know if the first child has been specified or not. We only know for certain after checking that second group and seeing if that was definitely set. If we treat CP<=1 as specifying a mutually exclusive group then this logic is impossible to fullfill... Because both groups will always be marked as specified (or completely fail if both params are set), and the parent exclusivity test will fail.

**Comment by nyalldawson** [comment]
(for reference : I had originally implemented the logic you propose. But the unit tests failed, and were impossible to get all passing until the fuzzy logic was added)

File: tests/src/analysis/testqgsprocessing.cpp (hunk 1 above)
Comment (nyalldawson) [reply tests/src/analysis/testqgsprocessing.cpp]: Yes - to me that's correct. It's saying you can optionally set the X and y resolution (the parent optional conditional), but if you specify one you HAVE to specify the other (the child all required condition)

Comment (nyalldawson) [reply tests/src/analysis/testqgsprocessing.cpp]: Yes - to me that's correct. It's saying you can optionally set the X and y resolution (the parent optional conditional), but if you specify one you HAVE to specify the other (the child all required condition)

**Comment by wonder-sk** [comment]
I think I am starting to understand where our concepts are different... when evaluating the groups, you are looking both at "is present" property of a group and at "is satisfied" property - and **making decisions by combining these two separate properties** which is where I am feeling lost in the logic.

My thinking is that there should be just one property - "is satisfied" - which would allow us to use just simple true/false logic. With the simpler logic we can implement all the group types except for `mutually_exclusive_optional` which I believe we can replace with something else.

Let's look again on the good example with resolution: we say that we want users to optionally set x/y resolution. That means we accept either {} or {X,Y} as valid inputs, but {X} or {Y} are invalid inputs.

In my understanding, with the simpler logic, the group should be defined as:

`mutually_exclusive_required(all_required(X,Y), none_required(X,Y))`.

(`none_required` being the opposite of `all_required` - satisfied when none of its children are satisfied)

Because this will be probably a common case that we require a bunch of parameters or none of them, I would suggest `all_or_none` type which would be a shorthand for `mutually_exclusive_required(all_required(a,b,c,d), none_required(a,b,c,d))`. It would also simplify our example with resolution to just `all_or_none(X,Y)`.

**Comment by nyalldawson** [comment]
@wonder-sk 

That could work - the missing bit was the (non obvious ;P) "none required" logic. Let me try and see if I can get this passing the checks.

**Comment by nyalldawson** [comment]
Actually on second thoughts - mutually exclusive optional is a really common requirement which we need to handle too.

**Comment by nyalldawson** [comment]
@wonder-sk I've struggled with adapting to your approach here, but my honest opinion is that while it simplifies the internal code, the external api becomes extremely complex.

E.g. for the past 20 minutes I've been trying to code a test for the following logic:

- Either the user has to enter both P1 and P2, or one of P3 or P4. 
- They can't enter P1 without P2 (and vice versa)

It becomes a difficult logic problem to solve just to encapsulate this in the grouping API. While I've probably overlooked a simple solution, it remains that if I'm struggling here, users of this API will have little chance of using it in a stress-free way!

For reference the previous approach allowed this to be specified via `at_least_one_required( all_required( p1 ,p2 ), mutually_exclusive_required( p3, p4 ) )`

**Comment by nyalldawson** [comment]
Closing pending a future redesign

**Pull request closed without merging**